sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/BatchingParDo.java

```diff
@@ -0,0 +1,214 @@
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.transforms;
+
+import com.google.common.annotations.VisibleForTesting;
+import org.apache.beam.sdk.coders.Coder;
+import org.apache.beam.sdk.coders.VarIntCoder;
+import org.apache.beam.sdk.coders.VarLongCoder;
+import org.apache.beam.sdk.transforms.windowing.BoundedWindow;
+import org.apache.beam.sdk.util.TimeDomain;
+import org.apache.beam.sdk.util.Timer;
+import org.apache.beam.sdk.util.TimerSpec;
+import org.apache.beam.sdk.util.TimerSpecs;
+import org.apache.beam.sdk.util.state.BagState;
+import org.apache.beam.sdk.util.state.StateSpec;
+import org.apache.beam.sdk.util.state.StateSpecs;
+import org.apache.beam.sdk.util.state.ValueState;
+import org.apache.beam.sdk.values.KV;
+import org.apache.beam.sdk.values.PCollection;
+import org.joda.time.Duration;
+import org.joda.time.Instant;
+import org.slf4j.LoggerFactory;
+
+/**
+ * {@link PTransform} that allows to compute elements in batch of desired size. The input {@link
+ * PCollection} needs to be a {@code PCollection<KV>}. Elements that must belong to the same batch
+ * need to have the same key. Elements are added to a buffer. When the buffer reaches {@code
+ * batchSize}, it is then processed through a user {@link SimpleFunction perBatchFn} function. The
+ * output elements then are added to the output {@link PCollection}. Windows are preserved (batches
+ * contain elements from the same window). Batching is done trans-bundles (batches may contain
+ * elements from more than one bundle)
+ *
+ * <p>Example (batch call a webservice and get return codes)
+ *
+ * <pre>{@code
+ * SimpleFunction<Iterable<String>, Iterable<String>> perBatchFn =
+ * new SimpleFunction<Iterable<String>, Iterable<String>>() {
+ *   {@literal @}Override
+ *   public Iterable<String> apply(Iterable<String> input) {
+ *     ArrayList<String> results = callWebService(input);
+ *     return results;
+ *     }
+ *  };
+ *  ...
+ *  Pipeline pipeline = Pipeline.create(...);
+ *  ...
+ *  long batchSize = 100L;
+ *  pipeline.apply(BatchingParDo.via(batchSize, perBatchFn))
+ *          .setCoder(StringUtf8Coder.of());
+ *  pipeline.run();
+ * }</pre>
+ * *
+ */
+public class BatchingParDo<K, InputT, OutputT>
+    extends PTransform<PCollection<KV<K, InputT>>, PCollection<KV<K, OutputT>>> {
+
+  private final long batchSize;
```

Comment:
I agree with the point on latency. 

Nevertheless the orientation of this feature was to provide exact batching (https://docs.google.com/document/d/1zf9TxIOsZf_fz86TGaiAQqdNI5OO7Sc6qFsxZlBAMiA/edit?usp=sharing) to allow use cases such as abuse detection in an auction system: raise an alert when a user has made exactly 10 auctions in the last hour. The state API allows to count to 10 even if a bundle is smaller than 10 elements. There are still corner cases in which, the exact batching cannot be obtained: if the window contains 15 elements, then there will be a batch of 10 elements and a batch of 5 elements because we need to output the batched elements in the same window as the input.

Regarding what you propose: if we flush batch at the end of the bundle, no need to use state then. And for the use of minBatchSize, it will be little often guaranteed: if after a batch flush, there is fewer than minBatchSize elements left in the window, then, to maintain windowing, we will output less than minBatchSize elements. That is why you said "targeted" I guess. Is this ok for a user to have min and max parameters that are not almost always guaranteed, or at least often?

Maybe it is at the end a compromise to find between possible use cases and latency?
WDYT?